Unified Course Team and Library Users page. Converted both to Backbone.

e-kolpakov opened the pull request:
**Background:** Library Users page was created by copy-pasting of Course Team due to Course Library MVP deadline. This PR contains a unification of two pages and switching to using Backbone and underscore
**Jira tickets:** [SOL-194](https://openedx.atlassian.net/browse/SOL-194), OC-45
**Discussions:** [Previous PR](https://github.com/edx/edx-platform/pull/6283/#issuecomment-67493618), [SOL-194](https://openedx.atlassian.net/browse/SOL-194)
**Sandbox:** [LMS](http://sandbox2.opencraft.com/),  [Studio](http://sandbox2.opencraft.com:18010/)
**Dependencies:** none
**Test instructions:** Library Users and Course Team pages should work as expected, namely:

1. Admin can add/delete, promote/demote users.
2. Non-admin users cannot add/delete, promote/demote other users (**changed behavior:** Course Users page did not allow deleting current user from course)
3. Users can delete themselves from course/library.
4. There must always be at least one admin, hence last admin cannot demote or delete himself (**changed behavior:** Course Team did not allow deleting current user; Library Users did not allow deleting current user if (s)he was last admin)
5. Course Team use two roles: 'staff' and 'instructor' (aka "Admin"); Library Users use three: 'library_user' (aka "User"), 'staff' and 'instructor' (aka "Admin").

**Author concerns:** new bok_choy tests might be flaky - one out of nine tests failed at `setUp` due to 404 in studio. Refreshing `page` object after every modification seems fixed the issue as page is refreshed each time, but the reason why only *one random* test failed out of nine is unclear.

edx-webhook commented:
Thanks for the pull request, @e-kolpakov! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?number=7336&repo=edx%2Fedx-platform

e-kolpakov commented:
@cahrens SOL-194 says that second reviewer would be on T&L team, so could you please assign someone to make a review?

e-kolpakov commented:
test this please

cahrens commented:
@e-kolpakov Why does this PR contain translation changes?

Please provide a link to the first review. I don't know if we will be able to review this PR in our current sprint, but I'll see what we can do.

e-kolpakov commented:
@cahrens First review will happen here.

There are some translation changes: some strings were moved from html templates to underscore, so  they moved from django messages to djangojs. If those files are not supposed to be included into PRs I'll remove them, but it looks like the mechanism that prevented them to be committed is a bit broken.

Also, is this "test this please" thing work? I was unable to get a build fired using this, but haven't heard anything about test team changing (or disabling) it.

There's no hurry with this, so take your time with reviewing. Thank you!

cahrens commented:
Please check with @sarina about how she wants to handle this translation update.

And if "test this please" isn't working for you, please check with testeng (@benpatterson). I don't know if the functionality is broken, or if you aren't on a list that allows is to work (being an external contributor). It has worked for me, but I haven't tried it recently.

sarina commented:
ok to test

sarina commented:
@e-kolpakov generally changes to .po/.mo files aren't checked in by developers. It causes a larger diff and means there's a burden to keep updating the files as you change them. I merge in updates to these files once per week, on Monday mornings, so you don't have to worry about them.

sarina commented on conf/locale/eo/LC_MESSAGES/django.po:
```diff
@@ -5045,6 +5104,18 @@ msgstr "Sérvér érrör. Pléäsé trý ägäïn ïn ä féw mïnütés. Ⱡ'σ
 
+#: lms/djangoapps/instructor/tests/test_api.py
+#: lms/djangoapps/instructor/views/api.py
+msgid "This student (%s) will skip the entrance exam."
```
Where do you define this string? Please use NAMED variables, not "%s". See http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/internationalization/i18n.html#general-internationalization-rules

e-kolpakov replied:
@sarina I don't. It's from other commit: https://github.com/edx/edx-platform/commit/e21ff20375e5c837b10ce7ad6c1340597266486f#diff-1dcea18a471b3b764091199f6d021480R2555

e-kolpakov commented:
@sarina ok, thanks for the explanation. BUt should I remove them from PR or leave as is?

sarina commented:
@e-kolpakov I would recommend removing them from your PR - if you generate them locally, it should be for testing purposes & you shouldn't check them in.

e-kolpakov commented:
@sarina removed.

benpatterson commented:
@e-kolpakov the test comment will only work if you are whitelisted. 

I'll add you to the whitelist now (only an admin can do this, which includes @sarina): add to whitelist

e-kolpakov commented:
@benpatterson understood, thank you.

Kelketek commented on cms/templates/js/mock/mock-manage-users-lib.underscore:
```diff
@@ -1,146 +1,57 @@
+        </article>
+      </section>
+    </div>
+</div>
```
Nit: Newline missing here.

Kelketek commented on cms/templates/js/team-member.underscore:
```diff
@@ -0,0 +1,50 @@
+    </ul>
+    <% } %>
+
+</li>
```
Nit: Newline missing here, too

Kelketek commented on common/test/acceptance/tests/studio/base_studio_test.py:
```diff
@@ -37,7 +37,7 @@ def populate_course_fixture(self, course_fixture):
         pass
 
-    def log_in(self, user, is_staff=False):
+    def log_in(self, user, is_staff=False, no_login=False):
```
This argument's presence is confusing. I'm still reading, so I'm sure you have a reason for it being here, but could you add something in the docstring explaining what it's for?

Kelketek commented on common/test/acceptance/tests/studio/test_studio_course.py:
```diff
@@ -0,0 +1,432 @@
+    """ As a course author, I want to be able to add others to my team """
+    def _make_user(self, username):
+        """ Registers user and returns user representation dictionary as expected by `log_in` function """
+        user = {
```
Why is this necessary?

e-kolpakov replied:
@Kelketek to register a user but do not log in. This is confusing, so I removed `no_login` parameter from `log_in` function and explicitly used `AutoAuthPage` instead

Kelketek commented on common/test/acceptance/pages/studio/users.py:
```diff
@@ -92,6 +136,30 @@ def url(self):
+        self.wait_for_ajax()
+        self.wait_for_element_invisibility(
+            '.ui-loading',
+            'Wait for the page to complete its initial loading of XBlocks via AJAX'
```
XBlocks? On the Team Course page? Are we sure this wait_until_ready function is doing what we intend for it to?

Kelketek commented:
@e-kolpakov The test failure is a little worrisome. I would expect to get a different error than a timeout if it were a 404 in this instance. `bok_choy`'s visit() function should only see if that URL loads at all, not that the page that's loaded is particularly ready. To me, this seems to indicate that the server didn't respond fast enough. But since the timeout is 60 seconds, I'd think it would have enough time to render the course team page.

It's possible Jenkins was doing some DNS lookup and got stuck while doing so or something, preventing the server from loading. But I'd like @jzoldak to weigh in. It seems odd that this test failure in particular is what hit us if that's the case, since the test is within the scope of the PR, and there are many others it could have happened to, statistically. Jay, is there a better explanation? Do I understand visit() correctly?

Kelketek commented:
@e-kolpakov Please don't forget to set up a sandbox for this. The team will want to verify things behave as expected when they click around.

e-kolpakov commented:
@Kelketek @jzoldak  I've investigated this thing a bit. Turns out, this is a very weird bug, and it's reasons still unknown to me. My observations:

* Any test out of those nine added can fail.
* Fail happens in `setUp` function
* It is always caused by 404 error.
* Failed page URL is: `http://localhost:8031/course_team/course-v1:test_org+2B336790434684799951055756418812500001273+test_run`
* If I manually reload the page it still shows 404
* If I manually go to dashboard the course with selected number exists.
* When I manually go to course team page for that course URL is **exactly the same as before**, but page opens and test continues to success.
* Tests that pass with no user interactions access course team page at `http://localhost:8031/course_team/course-v1%3Atest_org%2B336790434684799951055756418812500001273%2Btest_run, i.e. the same url with course id urlencoded.

I'm investigating it further. Meanwhile, I've addressed @Kelketek 's notes and will push an update in a minute.

e-kolpakov commented:
@Kelketek @jzoldak Update: urls are not quite the same; broken URL does not contain a trailing slash. Which means that for some reason URL normalizing (i.e. appending trailing slash if it's not present) does not work from time to time.

Kelketek commented:
@e-kolpakov Check the URL patterns. If Django does not detect that the URL pattern matches, it may (depending on configuration) append a slash internally before trying the route again, and finally giving a 404.

e-kolpakov commented:
@Kelketek actually it's quite the opposite: url with trailing slash works, without does not work. Since there's a `APPEND_SLASH` directive and it's set to True (implicitly: default is true and it's not overridden anywhere). It should append slash and try again. But for some reason it does not work from time to time.

URL pattern is the following: `r'^course_team/{}/(?P<email>.+)?$'.format(COURSELIKE_KEY_PATTERN)`, so trailing slash is required here. The only other pattern in the `cms/urls.py` that requires trailing slash is `r'^home/$'`.

This explains why there weren't any acceptance test issues before: this url was not tested using bok choy; and page url constructed differently in lettuce (always with trailing slash). So, since the common approach is to allow missing trailing slash, I'll update url pattern to be permissive as well.

Kelketek commented:
@e-kolpakov That's very strange. I wonder if the 404 was generated by it not finding a matching pattern, or by it matching something different than we expected, and raising a 404 when that function did not find what it expected. The latter might explain intermittent behavior if it's possible for URLs to be checked in a different order for some reason. I can't think of any cause for this off the top of my head, however, and your solution sounds like it may be the most practical anyway.

e-kolpakov commented:
@Kelketek url starts with `course_team` - it's unlikely to match anything different. But I'll take a look :)

edx-webhook commented:
Thanks for the pull request, @e-kolpakov! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?number=7336&repo=edx%2Fedx-platform

edx-webhook commented:
Thanks for the pull request, @e-kolpakov! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=7336

Kelketek commented:
:+1:

cahrens commented on common/djangoapps/terrain/steps.py:
```diff
@@ -233,3 +233,11 @@ def run_ipdb(_step):
     assert True
+
+
+@step(u'(I am viewing|s?he views) the course team settings$')
```
This is still used by a Lettuce test?

e-kolpakov replied:
@cahrens yes, cms/djangoapps/contentstore/features/help.feature

cahrens commented:
@e-kolpakov is this PR now ready for TNL review?

e-kolpakov commented:
@cahrens yes, I believe it is ready for TNL review

cahrens commented on cms/djangoapps/contentstore/views/user.py:
```diff
@@ -71,14 +81,20 @@ def _manage_users(request, course_key):
         'context_course': course_module,
-        'staff': staff,
-        'instructors': instructors,
+        'transfer_ownership': request.user in instructors and len(instructors) == 1,
```
This variable name confused me-- it isn't representing that transferring of ownership should/will happen, it's indicating that transferring of ownership can't happen and some text will be shown explaining the situation.

Perhaps "show_transfer_ownership_text" or something to that effect.

cahrens commented on cms/static/js/factories/manage_users.js:
```diff
@@ -1,203 +1,31 @@
+/*
+    Code for editing users and assigning roles within a course team context.
+*/
+define(['underscore', 'gettext', 'js/views/manage_users_and_roles', 'coffee/src/main'],
```
What is coffee/src/main needed for? Please add a comment since there is no way to tell from this file.

cahrens commented on cms/static/js/factories/manage_users_lib.js:
```diff
@@ -1,159 +1,35 @@
 */
-define(['jquery', 'underscore', 'gettext', 'js/views/feedback_prompt', 'js/views/utils/view_utils'],
-function($, _, gettext, PromptView, ViewUtils) {
+define(['underscore', 'gettext', 'js/views/manage_users_and_roles', 'coffee/src/main'],
```
Same comment/question about coffee/src/main.

cahrens commented on cms/static/js/spec/views/pages/library_users_spec.js:
```diff
@@ -9,11 +9,17 @@ function ($, AjaxHelpers, ViewHelpers, ManageUsersFactory, ViewUtils) {
+                ],
+                "dummy_change_role_url",
+                10000,
+                true
```
Are there any unit tests with allowActions=False?

cahrens commented on cms/static/js/views/manage_users_and_roles.js:
```diff
@@ -0,0 +1,306 @@
+            primaryAction: gettext('Return and add email address')
+        },
+        alreadyMember: {
+            title: gettext('Already a course team member'),
```
I found these "course team" specific messages confusing in this general file. You override it for both course teams and library teams, so perhaps don't supply it here (or make the text generic).

cahrens commented on cms/static/js/views/manage_users_and_roles.js:
```diff
@@ -0,0 +1,306 @@
+/*
+    Code for editing users and assigning roles within a course team context.
```
Update comment-- this is for both course teams and library teams.

cahrens commented on cms/static/js/views/manage_users_and_roles.js:
```diff
@@ -0,0 +1,306 @@
+            this.allow_actions = options.allow_actions;
+            this.current_user_id = options.current_user_id;
+
+            this.initial_role = this.roles[0];
```
I didn't realize that the ordering of the roles is important. A comment where the roles are created would be good.

cahrens commented on cms/static/js/views/manage_users_and_roles.js:
```diff
@@ -0,0 +1,306 @@
+        getPossibleRoleChangesForRole: function(role, adminRoleCount) {
+            var result = [],
+                role_names = _.map(this.roles, function(role) { return role.key });
+            if (role == this.admin_role.key && adminRoleCount==1) {
```
Nit: make spacing around operators consistent.

cahrens commented on cms/static/js/views/manage_users_and_roles.js:
```diff
@@ -0,0 +1,306 @@
+                // in reverse order to show "Add" buttons to the left, "Remove" to the right
+                for (var i=this.roles.length-1; i>=0; i--) {
+                    var other_role = this.roles[i];
+                    if (Math.abs(currentRoleIdx-i) != 1) continue; // allows moving only to adjacent roles
```
Use === and !== in JavaScript to avoid type coercion.

cahrens commented on common/test/acceptance/tests/studio/test_studio_course.py:
```diff
@@ -0,0 +1,436 @@
+
+
+@attr('shard_2')
+class CourseTeamPageTest(StudioCourseTest):
```
I would put the course team tests in their own file (test_studio_course_team.py) and move CoursePagesTest back to where it was.

cahrens commented on common/test/acceptance/tests/studio/base_studio_test.py:
```diff
@@ -42,6 +42,8 @@ def log_in(self, user, is_staff=False):
         Log in as the user that created the course. The user will be given instructor access
         to the course and enrolled in it. By default the user will not have staff access unless
         is_staff is passed as True.
+        :param dict user: dictionary containing user data: {'username': ..., 'email': ..., 'password': ...}
```
Nit: we use the Google documentation style for Python.

http://sphinxcontrib-napoleon.readthedocs.org/en/latest/example_google.html

cahrens commented on common/test/acceptance/tests/studio/test_studio_course.py:
```diff
@@ -0,0 +1,436 @@
+
+        actual_visible = any((check_course_equality(course, self.course_info) for course in courses))
+
+        self.assertTrue(actual_visible == visible)
```
Nit: can use assertEqual(actual, expected) which produces a somewhat better error message if it fails.

cahrens commented on common/test/acceptance/tests/studio/test_studio_course.py:
```diff
@@ -0,0 +1,436 @@
+        And I am viewing the course team settings
+        When I add other user to the course team
+        And I delete other user from the course team
+        And other user logs in
```
The test doesn't check these last 2 bits.

cahrens commented on common/test/acceptance/tests/studio/test_studio_course.py:
```diff
@@ -0,0 +1,436 @@
+        And I cannot manage users
+        And I cannot make myself a course team admin
+        """
+        self.page.add_user_to_course(self.other_user.get('email'))
```
Thank you for converting these Lettuce tests! I imagine you converted exactly what was there, but I'm finding the amount duplication between the tests to be a bit extreme. For instance, this test can easily be folded into the previous test by simply adding 

self._assert_can_not_manage_users()
self.assertFalse(current.can_promote)

after line 388. While I see the value in having independent tests for certain things, we also are constantly battling to make build times faster. Therefore I think it is worth losing the duplicative independent test cases in this case.

I would also delete test_admins_can_add_other_users, since the steps in it are the exact same first steps in test_added_users_cannot_add_or_delete_other_users.

cahrens commented on common/test/acceptance/pages/studio/users.py:
```diff
@@ -78,6 +80,48 @@ def click_submit_new_user_form(self):
+        return self.q(css='.prompt.{dialog_type} .message'.format(dialog_type=dialog_type)).text[0]
+
+
+class UsersPage(UsersPageMixin):
```
I don't think you need this separate UsersPage class anymore (it is only used by LibraryUsersPage).

cahrens commented on common/test/acceptance/pages/studio/users.py:
```diff
@@ -78,6 +80,48 @@ def click_submit_new_user_form(self):
+    def get_user(self, email):
+        """ Gets user wrapper by email """
+        target_users = [user for user in self.users if user.email == email]
+        assert len(target_users) == 1
```
Pages methods shouldn't do asserts. In this sort of method, we just return the first element.

e-kolpakov replied:
@cahrens the intention here was to emphasize that (1) it is not a test assertion and (2) there should be single user. I believe if for some reason there would be two users, assertion error thrown here would make more sense than possibly unwanted behavior if further actions are performed on wrong user.

cahrens commented on cms/urls.py:
```diff
@@ -75,13 +75,13 @@
     url(r'^request_course_creator$', 'request_course_creator'),
 
-    url(r'^course_team/{}/(?P<email>.+)?$'.format(COURSELIKE_KEY_PATTERN), 'course_team_handler'),
+    url(r'^course_team/{}(?:/(?P<email>.+))?$'.format(COURSELIKE_KEY_PATTERN), 'course_team_handler'),
```
For my own edification, why are the question marks important?

e-kolpakov replied:
@cahrens it's a non-capturing regex group. It tells regex engine that this group should not be remembered, i.e. it is only to apply "or" or quantity modifiers (`+`,`*`,`?`, `{}`). As a result, this group is not sent to django view.

cahrens replied:
If it is not sent to the Django view, how is the email used?

e-kolpakov replied:
@cahrens there are two groups: outermost is non-cpaturing, innermost is capturing and named. E.g. in `course_team/course_id/staff@example.com`, first (non-capturing) group would contain `/staff@example.com`, while second (capturing, named) - `staff@example.com`.

The reason behind this is to make entire part after course_id optional, but still capturing email and sending it to view.

cahrens commented on common/test/acceptance/pages/studio/users.py:
```diff
@@ -92,6 +136,29 @@ def url(self):
+
+    url_path = "course_team"
+
+    def wait_until_ready(self):
```
Isn't this also needed for the library page?

cahrens commented on cms/templates/js/mock/mock-manage-users-lib.underscore:
```diff
@@ -1,146 +1,57 @@
+          <div class="wrapper-create-element animate wrapper-create-user">
+            <form class="form-create create-user" id="create-user-form" name="create-user-form">
+              <div class="wrapper-form">
+                <h3 class="title">Grant Instructor Access to This Library</h3>
```
It would have been nice if this "add user" section of the page was also in Backbone. I won't hold up the PR for it, but did you consider it?

e-kolpakov replied:
@cahrens yes, I did. The reason not to do this was that I saw it as a part of much larger scope, i.e. don't use `allow_actions` conditional rendering in django template. And that was a bit more involving than I could take in a single leap.

cahrens commented on cms/static/js/views/manage_users_and_roles.js:
```diff
@@ -0,0 +1,306 @@
+                type: newRole ? 'POST' : 'DELETE',
+                dataType: 'json',
+                contentType: 'application/json',
+                notifyOnError: false,
```
What does setting notifyOnError to false here do? Does it prevent the general error handling from taking over?

e-kolpakov replied:
@cahrens taken as-is from where it lived earlier.

cahrens commented on cms/static/js/factories/manage_users.js:
```diff
@@ -1,203 +1,31 @@
+        }
+
+        var options = {
+            el: $("#content"),
```
Is this the general "#content" element from base.html? With Backbone, you really should only give the backbone view the element that it is responsible for rendering. Can you scope this more tightly? I think part of the issue is that the add users form is not part of the Backbone view (though I would like it to be).

e-kolpakov replied:
@cahrens this view binds "Add New User" event, and events only bound to elements under `el`. So it needs to wrap both actual content and navbar. Unfortunately, `#content` is a closest common parent for navbar and content.

cahrens commented on cms/static/js/views/manage_users_and_roles.js:
```diff
@@ -0,0 +1,306 @@
+        },
+
+        keyUpHandler: function(event) {
+            if(event.which == jQuery.ui.keyCode.ESCAPE && this.$createUserFormWrapper.is('.is-shown')) {
```
===

cahrens commented:
Done reviewing-- nice work! The only thing I'm squeamish about his the fact that part of the page (in particular, the add user form) is still being rendered server-side, increasing the amount of "reaching out" that the Backbone view has to do for handling events. We have commonly left that special "new button" on the RHS out of the Backbone code (which does lead to a bit of ugliness), but I'd love to see the new user form itself moved to Backbone.

e-kolpakov commented:
@cahrens I've addressed minor notes, but I would like to discuss "convert new user form to Backbone". Do you think this PR is good enough with new user form still rendering server-side? I would like to do this conversion in smaller chunks if possible. I mean, converting the rest of the page to backbone would be a separate PR later (likely with dedicated task to allow for scheduling/prioritizing).

Also, I have no better commit message for last commit than "Addressed review notes", so I promise I'll squash commits when we're done with review.

cahrens commented on cms/static/js/spec/views/pages/library_users_spec.js:
```diff
@@ -5,76 +5,151 @@ define([
+                expect($('.wrapper-create-user')).not.toHaveClass('is-shown');
+            });
+
+            // Disabled flaky test - the following three disabled by bradenmacdonald on 2015-02-09
```
Remove comment.

cahrens commented on common/djangoapps/terrain/steps.py:
```diff
@@ -233,3 +233,19 @@ def run_ipdb(_step):
+
+
+@step(u'(I am viewing|s?he views) the course team settings$')
+def view_course_team_settings(_step, whom):
```
Duplicated method.

cahrens commented:
:+1: after two minor issues noted above are fixed. Thanks for the good work!

@e-kolpakov I agree with handling the new user form as a separate task.
